network/transport/grpc/connection_manager.go

```diff
@@ -213,16 +264,103 @@ func (s *grpcConnectionManager) Stop() {
+func (s *grpcConnectionManager) dialerLoop() {
+	log.Logger().Info("start dialing")
+	ticker := time.NewTicker(time.Second)
+	dialingWG := new(sync.WaitGroup)
+	defer ticker.Stop()
+outerLoop:
+	for {
+		select {
+		case <-s.ctx.Done():
+			break outerLoop
+		case <-ticker.C:
+			for _, c := range s.addressBook.limit(maxConcurrentDialers, isNotActivePredicate(s), backoffExpiredPredicate(), notDialingPredicate()) {
+				// the notDialingPredicate above guarantees that dialing is currently false. We can take the dialing lock
+				c.dialing.Store(true)
+				dialingWG.Add(1)
+				go func(cp *contact) {
+					defer func() {
+						cp.dialing.Store(false) // reset call lock at the end of dialing
+						dialingWG.Done()
+					}()
+					s.dial(cp)
+				}(c)
+			}
+		}
+	}
+	dialingWG.Wait()
+}
+
+func (s *grpcConnectionManager) dial(contact *contact) {
+	connection, isNew := s.connections.getOrRegister(s.ctx, contact.peer, true)
 	if !isNew {
-		log.Logger().
-			WithField(core.LogFieldPeerAddr, peer.Address).
-			Info("Connection for peer already exists.")
+		// can only occur when receiving an inbound connection at the same time.
+		log.Logger().WithFields(contact.peer.ToFields()).
+			Debug("stop dialing, already has a connection")
+		return
+	}
+	defer func() {
+		// connection does not exist outside the dialer
+		connection.disconnect()
+		s.connections.remove(connection)
+	}()
+
+	// Open a grpc.ClientConn
+	log.Logger().WithFields(contact.peer.ToFields()).Debug("connecting to peer")
+	contact.attempts.Add(1)
+	contact.lastAttempt.Store(time.Now())
+	dialContext, cancel := context.WithTimeout(s.ctx, s.connectionTimeout)
+	defer cancel()
+	grpcClient, err := s.dialer(dialContext, contact.peer.Address, s.dialOptions...)
+	if err != nil { // failed to connect
+		log.Logger().WithError(err).WithFields(contact.peer.ToFields()).Debug("failed to open a grpc ClientConn")
+		errStatus, isStatusError := status.FromError(err)
+		if isStatusError && errStatus.Code() == codes.Canceled {
+			// Do not backoff when context is cancelled
+			// Backoff might try to persist after stores are closed
+			// https://github.com/nuts-foundation/nuts-node/issues/1864
+			return
+		}
+		contact.backoff.Backoff() // backoff store
 		return
 	}
-	s.startTracking(peer.Address, connection)
+	defer grpcClient.Close()
+	log.Logger().WithFields(contact.peer.ToFields()).Debug("connected to peer (outbound)")
+
+	// Connect protocol streams
+	err = s.openOutboundStreams(connection, grpcClient) // blocking call, dial needs to be async
+	if err != nil {
+		// connection failed, increase backoff
+		// TODO: check if this works as intended for multiple streams/protocols on the same connection
+		contact.backoff.Backoff()
+		if errors.Is(err, ErrUnexpectedNodeDID) {
+			contact.backoff.Reset(time.Hour * 24 * 365 * 100) // backoff expires after 100 years. Backoff must be reset when peer's (DID) address is updated.
```

Comment:
this most likely happens when a NutsComm address is reused by a new DID. We don't expect the old DID to come back, so we just block it for a very long time or until the DID's NutsComm endpoint is updated.

This does not prevent the other node from calling us, so we could also reset the backoff after a successful inbound call?

This does not solve the case where node without a DID/NutsComm endpoint has blocked the contact since this node will not receive an inbound call.

Suggestions?

Reply:
Since it's a theoretical problem, for now make sure it's in the diagnostics and there's a way to reset backoff if required? (or maybe make an issue for the last feature ('good first issue'))?